tools/heapsnapshot.py

```diff
@@ -0,0 +1,217 @@
+
+    def getStringSnapshotIndex(string):
+        assert isinstance(string, str), f"'{string}' is not of type string"
+        if string not in stringToSnapshotIndex:
+            strings.append(string)
+            stringToSnapshotIndex[string] = len(strings) - 1
+        return stringToSnapshotIndex[string]
+
+    def getNodeSnapshotIndex(address):
+        # This is the index of the first element of the node in the flattened nodes array
+        return addressToNodeIndex[address] * nodeFieldCount
+
+    for address in addresses:
+        obj = data["objects"][address]
+        edgeCount = 0
+
+        if obj["type"] == "table":
+            name = f"Registry ({address})" if address == data["roots"]["registry"] else f"Luau table ({address})"
+            if "pairs" in obj:
+                for i in range(0, len(obj["pairs"]), 2):
+                    key = obj["pairs"][i]
+                    value = obj["pairs"][i + 1]
+                    if key is None and value is None:
+                        # Both the key and value are value types, nothing meaningful to add here
+                        continue
+                    elif key is None:
+                        edgeCount += 1
+                        addEdge([edgeTypeToMetaIndex["property"], getStringSnapshotIndex("Luau table key (value type)"), getNodeSnapshotIndex(value)])
+                    elif data["objects"][key]["type"] == "string":
+                        edgeCount += 1
+                        # This is a special case where the key is a string, so we can use it as the property name
+                        addEdge([edgeTypeToMetaIndex["property"], getStringSnapshotIndex(data["objects"][key]["data"]), getNodeSnapshotIndex(value if value is not None else key)])
+                    elif value is None:
+                        edgeCount += 1
+                        addEdge([edgeTypeToMetaIndex["property"], getStringSnapshotIndex(f'{data["objects"][key]["type"]} ({key})'), getNodeSnapshotIndex(key)])
+                    else:
+                        edgeCount += 2
+                        addEdge([edgeTypeToMetaIndex["property"], getStringSnapshotIndex(f'{data["objects"][key]["type"]} ({key})'), getNodeSnapshotIndex(value)])
+                        addEdge([edgeTypeToMetaIndex["internal"], getStringSnapshotIndex(f'{data["objects"][key]["type"]} ({key})'), getNodeSnapshotIndex(key)])
+            if "array" in obj:
+                for i, element in enumerate(obj["array"]):
+                    edgeCount += 1
+                    addEdge([edgeTypeToMetaIndex["element"], i, getNodeSnapshotIndex(element)])
+            if "metatable" in obj:
+                edgeCount += 1
+                addEdge([edgeTypeToMetaIndex["internal"], getStringSnapshotIndex(f'metatable ({obj["metatable"]})'), getNodeSnapshotIndex(obj["metatable"])])
+            addNode([nodeTypeToMetaIndex["object"], getStringSnapshotIndex(name), getUniqueId(address), obj["size"], edgeCount, 0, 0])
```

Comment:
I was a bit lazy here wrt detecting "arrays" vs "objects" and just called them all objects. I'm not sure if there's much value in distinguishing them, but I'm open to reworking this to accommodate that if we want to

Reply:
I added a todo for now